interfaces/connection.go

```diff
@@ -49,6 +50,38 @@ type ConnectedSlot struct {
 	Attrs map[string]interface{}
 }
 
+// Attrer is an interface with Attr getter method common
+// to ConnectedSlot, ConnectedPlug, PlugInfo and SlotInfo types.
+type Attrer interface {
+	Attr(key string, val interface{}) error
+}
+
+func getAttribute(snapName string, ifaceName string, staticAttrs map[string]interface{}, dynamicAttrs map[string]interface{}, key string, val interface{}) error {
+	var v interface{}
+	var ok bool
+
+	v, ok = dynamicAttrs[key]
+	if !ok {
+		v, ok = staticAttrs[key]
+	}
+
+	if !ok {
+		return fmt.Errorf("snap %q does not have attribute %q for interface %q", snapName, key, ifaceName)
+	}
+
+	rt := reflect.TypeOf(val)
+	if rt.Kind() != reflect.Ptr || val == nil {
+		return fmt.Errorf("internal error: cannot get %q attribute of interface %q with non-pointer value", key, ifaceName)
+	}
+
+	if reflect.TypeOf(v) != rt.Elem() {
+		return fmt.Errorf("snap %q has interface %q with invalid value type for %q attribute", snapName, ifaceName, key)
```

Comment:
I find this error a bit too abstract,

"snap %q has interface %q with %q attribute value of invalid type %v, expected %v", snapName, ifaceName, key, reflect.TypeOf(v), rt.Elem()

Reply:
I included type names initially but changed this per earlier request from Gustavo: 

"Similarly, this message is going to surface to users due to errors from snap publishers. Neither of those need to be Go developers, so we don't want to expose strings such as *map[string]interface{} in an error message.

Also, this is lacking some context that would enable to user or publisher to tell where that attribute is.

Needs to be something similar to:

"snap %q has interface %q with invalid value type for %q attribute"

Reply:
but 'invalid value type' is not very user friendly either, anyway my worry is that we will miss debug info, maybe we should log the message with types, and surface the one without:

"snap %q has interface %q with %q attribute value with unexpected type"

Reply:
Fair enough, I'll address this in a follow up PR. Thanks.